docs/extending_fuel.rst

```diff
@@ -0,0 +1,268 @@
+method, which apply on an individual source.
+
+There also exists an :class:`AgnosticSourcewiseTransformer` class for cases
+where the example and batch implementations of a sourcewise transformer are the
+same. This class requires a :meth:`transform_any_source` method to be
+implemented.
+
+>>> from fuel.transformers import AgnosticSourcewiseTransformer
+>>> 
+>>> class Doubler(AgnosticSourcewiseTransformer):
+...     def __init__(self, data_stream, **kwargs):
+...         super(Doubler, self).__init__(
+...             data_stream=data_stream,
+...             produces_examples=data_stream.produces_examples,
+...             **kwargs)
+... 
+...     def transform_any_source(self, source, _):
+...         return 2 * source
+
+Let's try this implementation on our dummy dataset.
+
+>>> target_stream = Doubler(
+...     data_stream=DataStream(
+...         dataset=dataset,
+...         iteration_scheme=batch_scheme),
+...     which_sources=('targets',))
+>>> all_stream = Doubler(
+...     data_stream=DataStream(
+...         dataset=dataset,
+...         iteration_scheme=batch_scheme),
+...     which_sources=None)
+>>> print([batch for batch in target_stream.get_epoch_iterator()])
+[(array([1, 2]), array([-2,  2])), (array([3, 4]), array([-2,  2]))]
+>>> print([batch for batch in all_stream.get_epoch_iterator()])
+[(array([2, 4]), array([-2,  2])), (array([6, 8]), array([-2,  2]))]
+
+Finally, there exists a :class:`Mapping` transformer which acts as a swiss-army
+knife transformer. In addition to a data stream, its constructor accepts a
+function which will be applied to data coming from the stream.
+
+Here's how you would implement the feature doubler using :class:`Mapping`.
+
+>>> from fuel.transformers import Mapping
+>>> 
+>>> features_index = dataset.sources.index('features')
+>>> def double(data):
+...     data = list(data)
+...     data[features_index] *= 2
+...     return tuple(data)
+>>> mapping_stream = Mapping(
+...     data_stream=DataStream(
+...         dataset=dataset, iteration_scheme=batch_scheme),
+...     mapping=double)
+>>> print([batch for batch in mapping_stream.get_epoch_iterator()])
+[(array([2, 4]), array([-1,  1])), (array([6, 8]), array([-1,  1]))]
+
+
+New iteration schemes
+---------------------
+
+New iteration schemes are implemented by subclassing :class:`IterationScheme`
+and implementing a :meth:`get_request_iterator` method, which should return an
+iterator that returns lists of indices.
+
+Two subclasses of :class:`IterationScheme` typically serve as a basis for other
+iteration schemes: :class:`IndexScheme` (for schemes requesting examples) and
+:class:`BatchScheme` (for schemes requesting batches). Both subclasses are
+instantiated by providing a list of indices or a number of examples, and
+:class:`BatchScheme` accepts an additional ``batch_size`` argument.
+
+Here's how you would implement an iteration scheme that iterates over even
+examples:
```

Comment:
The `imap` and `partition_all` function might be non-trivial for an entry-level python user. Maybe you could elaborate on what is returned (without getting into the details of `imap` and `partition_all` as they are not the core of this tutorial).